[ambientweather] Ambient Weather Binding initial contribution

mhilbush opened the pull request:
Supports retrieval of personal weather station data from the Ambient Weather service using their socket.io-based real-time API.

The following weather station models are supported in the initial release. Other models can be added relatively easily.
* WS-0900-IP
* WS-1400-IP (Observer IP-based, which also includes the WS-1401-IP)
* WS-2902A
* WS-8482

Signed-off-by: Mark Hilbush <mark@hilbush.com>

This is ready for review.

mhilbush commented:
@wborn Thanks for doing the legwork on testing the updated socket.io libs for openhabcloud. Made me much more comfortable updating my binding to those same libs. 👍

wborn commented:
That's a nice bonus indeed! The release notes showed there were lots of bug fixes. :-)

davidgraeff commented on addons/binding/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/json/EventDataJson.java:
```diff
@@ -0,0 +1,293 @@
+ *
+ * @author Mark Hilbush - Initial Contribution
+ */
+public class EventDataJson {
```
Can you please rename the json package to "model" or "dto"? That is the established name for data transfer objects.

mhilbush replied:
Done

davidgraeff commented on addons/binding/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,238 @@
+
+    public AmbientWeatherBridgeHandler(Bridge bridge) {
+        super(bridge);
+        gson = new Gson();
```
Can you please assign this straight on the field

mhilbush replied:
Done

davidgraeff commented on addons/binding/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherEventListener.java:
```diff
@@ -0,0 +1,351 @@
+    private AmbientWeatherBridgeHandler bridgeHandler;
+
+    public AmbientWeatherEventListener(AmbientWeatherBridgeHandler bridgeHandler) {
+        gson = new Gson();
```
Please get the json instance from the handler. It is best to keep only one instance per OSGi service (which in this case is your factory, I think).

mhilbush replied:
Reworked to reduce number of Gson instances, as well as avoid instantiating on each call to process data.

mhilbush commented:
Thanks for the review @davidgraeff! I'll work on resolving your feedback.

I've already migrated a few of my bindings, but was holding off on this one until the 3rd party jar issue was sorted out. With that now resolved, I'll include the bnd migration in my changes.

mhilbush commented:
BTW, speaking of Gson, I think I came across an issue while migrating another binding that uses Gson. I recall seeing somewhere that the Gson version included in the core was changed. And, I think that change breaks the ability to run newly built bindings on 2.5 SNAPSHOT versions prior to the Gson version change.

I don't know if there's anything that can/should be done about this, other than an awareness that you'll generate an unresolved requirement trying to run a newly built binding in an older 2.5 SNAPSHOT build.

```
2019-03-23 10:44:54.568 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/opt/openhab2/addons/org.openhab.binding.bhyve-2.5.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.binding.bhyve [275]
  Unresolved requirement: Import-Package: com.google.gson; version="[2.8.0,3.0.0)"
```

davidgraeff commented:
We don't need to care about snapshot users fortunately. But the back migration of ESH made Markus, Kai and the others to update a lot of libraries (gson, xtend, etc). I'm pretty sure that OH 2.4 bindings will be incompatible to OH 2.5-release (even the OSGi manifests are created differently now).

So it's not really an issue in my opinion.

mhilbush commented:
Pushed one commit to address the review feedback. I'll push a separate commit later for the bnd migration. That may not be for a few days as I'll be traveling.

mhilbush commented:
@davidgraeff I pushed a commit to convert to the new build system. That should address the last of your review feedback. Let me know if there's anything else that needs attention. Your help with this is much appreciated!

davidgraeff commented:
There are a lot of migrations happen at the moment, so we run into conflicts.
Could you solve them when you have time?

mhilbush commented:
I'll fix the merge conflicts. There are a couple other things I see that are wrong as well, so I need to fix those.

mhilbush commented:
Ok, @davidgraeff, I think I got it all sorted out.

davidgraeff commented:
Just wanted to merge and I noticed a conflict again :/

mhilbush commented:
@davidgraeff  I fixed the conflict. Please merge quickly before another one pops up. 😄

mhilbush commented:
Ack. I fixed the conflict using the Github UI, and now my last commit is not signed. Can I fix that using the Github UI?

mhilbush commented:
@davidgraeff 

Do I need to fix this, or can it be merged without the sign-off on the last commit? That **Resolve Conflicts** button was just too tempting, but I didn't realize it wouldn't let me add the Signed-off-by to the commit.

davidgraeff commented:
All good. I face the same problem all the time.

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherEventListener.java:
```diff
@@ -0,0 +1,351 @@
+    private String applicationKey;
+
+    // Socket.io socket used to access Ambient Weather real-time API
+    private Socket socket;
```
Not so cool actually. We have a websocket factory and a http factory in the framework, not very complicated to use. Is there a reason for socket.io here?

mhilbush replied:
As mentioned in my comment below, I'm using the Jetty websocket client in another binding I'm working on, and it's working just fine. I spent quite a few hours trying to get the Jetty websocket client to work with the Ambient Weather backend, but was unable to make it work (was unable to make a successful connection with the backend service). That was the reason for switching to socklet.io. For the socket.io libs, I tried to stay in sync with the ones openhabcloud is using.

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/model/DevicesJson.java:
```diff
@@ -0,0 +1,47 @@
+ * @author Mark Hilbush - Initial Contribution
+ */
+public class DevicesJson extends ArrayList<DevicesJson.Container> {
+    public class Container {
```
public static class.

and: Does it need to be an ArrayList, or does a Collection or List suffice? Generally people here don't like extending a java container. Not so sure if that is true for dto's as well. It is also not necessary. gson allows to (de)serialize to/from a `List<Container>`

mhilbush replied:
Done. I think I found an even simpler solution. Not sure why I missed that the first time. 😕

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/processor/AbstractProcessor.java:
```diff
@@ -0,0 +1,156 @@
+     */
+    public String convertWindDirectionToString(double windDirectionDegrees) {
+        String result = "UNKNOWN";
+        if (windDirectionDegrees >= 0 && windDirectionDegrees <= 11.25) {
```
Can you make add those entries into a `Treemap` please. See https://docs.oracle.com/javase/6/docs/api/java/util/TreeMap.html#ceilingKey(K). 

A data structure will not only be faster but also more easy to extend or adapt in the future (that is of course not necessary in this case. It's still faster)

mhilbush replied:
@davidgraeff 

I found an approach used in another binding ([synopanalyzer](https://github.com/openhab/openhab2-addons/blob/1fd57c40758fde6f8eca936f285ef0bf06894c79/bundles/org.openhab.binding.synopanalyzer/src/main/java/org/openhab/binding/synopanalyzer/internal/handler/SynopAnalyzerHandler.java#L179)). While it doesn't use a TreeMap, it does a simple index of an array (albeit at the cost of some upfront math). It does need some minor modification to avoid a potential AIOOB exception. WDYT?
```
private String getWindDirection(int degree) {
    double step = 360.0 / WIND_DIRECTIONS.length;
    double b = Math.floor((degree + (step / 2.0)) / step);
    return WIND_DIRECTIONS[(int) (b % WIND_DIRECTIONS.length)];
}

public static final String[] WIND_DIRECTIONS = new String[] { "N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE", "S",
        "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW" };
```

davidgraeff replied:
That's fine as well, the data is continuous so an array with computing the index should work. The modulo makes sure that we do not get out of bounds exceptions, doesn't it?

mhilbush replied:
> That's fine as well,

Ok, thanks.

> The modulo makes sure that we do not get out of bounds exceptions, doesn't it?

It sure does!

mhilbush replied:
BTW, to use this piece of code, do I need to add an Also-by to the PR and/or commit?

davidgraeff replied:
Nope. Not because of 5 lines of code without any special algorithm inside

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/processor/ProcessorFactory.java:
```diff
@@ -0,0 +1,78 @@
+    private static final Ws1400ipProcessor WS1400IP_PROCESSOR = new Ws1400ipProcessor();
+    private static final Ws2902aProcessor WS2902A_PROCESSOR = new Ws2902aProcessor();
+    private static final Ws8482Processor WS8482_PROCESSOR = new Ws8482Processor();
+    private static final Ws0900ipProcessor WS0900IP_PROCESSOR = new Ws0900ipProcessor();
```
This is not the typical factory pattern. Why are those types already created and not created on demand? If you have only one of those types, the other ones are still created and occupy ram.

mhilbush replied:
Reworked to create processor instances only when required.

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/processor/RemoteSensor.java:
```diff
@@ -0,0 +1,130 @@
+     */
+    private String convertSoilMoistureToString(double soilMoisture) {
+        String result = "UNKNOWN";
+        if (soilMoisture >= 0 && soilMoisture < 33.0) {
```
Data structure

mhilbush replied:
Done. I used a TreeMap for this. It works well but seems a bit overkill for the small number of values. Maybe I didn't implement it right. Will look forward to your feedback on this.

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/processor/Ws0900ipProcessor.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public void processWeatherData(AmbientWeatherStationHandler handler, String station, String jsonData) {
+        try {
+            logger.debug("Station {}: Updating weather data channels", station);
```
Try to make try-catch blocks as short as possible. Let it only wrap the json part and do the channel updates after

mhilbush replied:
Done for all processors.

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/processor/Ws1400ipProcessor.java:
```diff
@@ -0,0 +1,126 @@
+    @Override
+    public void processInfoUpdate(AmbientWeatherStationHandler handler, String station, String name, String location) {
+        // Update name and location channels
+        logger.debug("Station {}: Updating station information channels", station);
```
Don't log infront of framework update calls. They are already logged

mhilbush replied:
Done for all processors.

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/processor/Ws2902aProcessor.java:
```diff
@@ -0,0 +1,118 @@
+    @Override
+    public void processWeatherData(AmbientWeatherStationHandler handler, String station, String jsonData) {
+        try {
+            logger.debug("Station {}: Updating weather data channels", station);
```
See note from above

mhilbush replied:
I revised this debug message to make it more accurately reflect what is happening.

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/util/SlidingTimeWindow.java:
```diff
@@ -0,0 +1,55 @@
+@NonNullByDefault
+class SlidingTimeWindow<T> {
+    protected long period = 0;
+    protected final SortedMap<Long, T> storage = Collections.synchronizedSortedMap(new TreeMap<>());
```
The field can be a "Map" as well if you don't use any of the SortedMap specific functions (and there are none afaik)

mhilbush replied:
I'm actually using the firstKey and lastKey methods of the SortedMap class.

mhilbush commented:
> I have left some inline comments for this second review round.

Thanks. I'll work on the changes to address your feedback.

> You have migrated the binding, nice.

Thanks.

> But can you also remove all "libs" and use maven dependencies instead? You would need to reference those in the karaf feature file as well, have a look for examples in there.

Sure, but I might need some guidance. I actually tried to do this a while back, but ran into trouble trying to test the binding when I dropped it into the addons directory. I'll give it another go.

> You are using socket.io. Is there a reason? It is not a problem in general, just harder to maintain in the future with additional libraries.

I'm using the Jetty websocket client in [another binding](https://github.com/mhilbush/openhab2-addons/blob/bhyve-binding/bundles/org.openhab.binding.bhyve/src/main/java/org/openhab/binding/bhyve/internal/handler/EventListener.java) I'm working on. It's working well. But, I spent quite a few hours trying to get it to work with the Ambient Weather backend, but was unable to make it work. That was the reason for switching to socklet.io. For the socket.io libs, I tried to stay in sync with the ones openhabcloud is using.

> This contribution is over 3k lines, because there are a lot of static data. Have a look at #5337.
Would it be feasible for you to move some static data that is expressed in code at the moment into a bundle resource file and create data structures during runtime?

Sure.

davidgraeff commented:
> Sure, but I might need some guidance. I actually tried to do this a while back, but ran into trouble trying to test the binding when I dropped it into the addons directory. I'll give it another go.

You need to drop the dependencies to the addon directory as well. But that's what we want to archive here, that the generated bundles are no longer "fat".

mhilbush commented:
> You need to drop the dependencies to the addon directory as well.

Ah, that definitely was the step I missed. :wink: Thanks!

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/config/BridgeConfig.java:
```diff
@@ -0,0 +1,64 @@
+ *
+ * @author Mark Hilbush - Initial contribution
+ */
+public class BridgeConfig {
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/config/StationConfig.java:
```diff
@@ -0,0 +1,48 @@
+ *
+ * @author Mark Hilbush - Initial contribution
+ */
+public class StationConfig {
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,237 @@
+    private final Logger logger = LoggerFactory.getLogger(AmbientWeatherBridgeHandler.class);
+
+    // URL to retrieve device list from Ambient Weather
+    private static String devicesUrl = "https://api.ambientweather.net/v1/devices?applicationKey=%APPKEY%&apiKey=%APIKEY%";
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,237 @@
+
+    @Override
+    public void initialize() {
+        logger.debug("Bridge: Initializing ambientweather bridge handler.");
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,237 @@
+    private boolean haveApplicationKey() {
+        String configApplicationKey = getConfigAs(BridgeConfig.class).getApplicationKey();
+        if (StringUtils.isEmpty(configApplicationKey)) {
+            logger.debug("Bridge: Application key is missing for thing {}", getThing().getUID());
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,237 @@
+    private boolean haveApiKey() {
+        String configApiKey = getConfigAs(BridgeConfig.class).getApiKey();
+        if (StringUtils.isEmpty(configApiKey)) {
+            logger.debug("Bridge: API key is missing for thing {}", getThing().getUID());
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    public void updateThingStatus(String errorDetail, String statusDescription) {
+        logger.debug("Bridge: Key validation FAILED. Setting bridge OFFLINE: {}", errorDetail);
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,237 @@
+
+    @Override
+    public void dispose() {
+        logger.debug("Bridge: Disposing ambientweather bridge handler.");
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherEventListener.java:
```diff
@@ -0,0 +1,351 @@
+public class AmbientWeatherEventListener {
+    private final Logger logger = LoggerFactory.getLogger(AmbientWeatherEventListener.class);
+
+    private static String realtimeUrl = "https://api.ambientweather.net/?api=1&applicationKey=%APPKEY%";
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherEventListener.java:
```diff
@@ -0,0 +1,351 @@
+    private final ConcurrentHashMap<AmbientWeatherStationHandler, String> handlers = new ConcurrentHashMap<>();
+
+    // URL used to subscribe or unsubscribe from weather data events
+    private final String subUnsubJson = "{ apiKeys: [ '%APIKEY%' ] }";
```
Done

mhilbush commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherEventListener.java:
```diff
@@ -0,0 +1,351 @@
+    private static String realtimeUrl = "https://api.ambientweather.net/?api=1&applicationKey=%APPKEY%";
+
+    // Maintain mapping of handler and weather station MAC address
+    private final ConcurrentHashMap<AmbientWeatherStationHandler, String> handlers = new ConcurrentHashMap<>();
```
Done

mhilbush commented:
@davidgraeff 

I've completed all the changes except for the conversion from local libs to maven dependencies. I'm still running into trouble trying to get the binding to run by placing it in the `addons` directory.

What I've done so far:
- removed `.classpath` entries for the local libs
- remove the `libs` directory
- added dependencies to `pom.xml`
- placed the dependent jars in the `addons` directory

I ran `mvn clean install`. The build was successful, and I can see where the libs were downloaded into my local maven repo.

But, when I drop the binding jar into addons, it still complains about the unresolved requirement.
```
2019-04-04 15:38:32.998 [DEBUG] [org.openhab.binding.ambientweather                ] - BundleEvent UPDATED - org.openhab.binding.ambientweather
2019-04-04 15:52:41.001 [ERROR] [org.openhab.binding.ambientweather                ] - FrameworkEvent ERROR - org.openhab.binding.ambientweather
org.osgi.framework.BundleException: Could not resolve module: org.openhab.binding.ambientweather [201]
  Unresolved requirement: Import-Package: io.socket.client
```

I didn't change feature.xml yet, because I didn't think that would matter when running the binding from the addons directory.

What did I miss?

Here's what I put in pom.xml
```
	<dependencies>
		<!-- https://mvnrepository.com/artifact/io.socket/socket.io-client -->
		<dependency>
			<groupId>io.socket</groupId>
			<artifactId>socket.io-client</artifactId>
			<version>1.0.0</version>
		</dependency>

		<!-- https://mvnrepository.com/artifact/com.squareup.okhttp3/okhttp -->
		<dependency>
			<groupId>com.squareup.okhttp3</groupId>
			<artifactId>okhttp</artifactId>
			<version>3.12.1</version>
		</dependency>

		<!-- https://mvnrepository.com/artifact/com.squareup.okio/okio -->
		<dependency>
			<groupId>com.squareup.okio</groupId>
			<artifactId>okio</artifactId>
			<version>1.15.0</version>
		</dependency>

		<!-- https://mvnrepository.com/artifact/io.socket/engine.io-client -->
		<dependency>
			<groupId>io.socket</groupId>
			<artifactId>engine.io-client</artifactId>
			<version>1.0.0</version>
		</dependency>

		<!-- https://mvnrepository.com/artifact/org.json/json -->
		<dependency>
			<groupId>org.json</groupId>
			<artifactId>json</artifactId>
			<version>20180813</version>
		</dependency>
	</dependencies>
```

davidgraeff commented:
You forgot to add `<scope>provided</scope>` to each dependency entry

mhilbush commented:
Hmm. I did that and it's still not resolving.

Added it on all the deps in the pom. Like this...
```
		<dependency>
			<groupId>io.socket</groupId>
			<artifactId>socket.io-client</artifactId>
			<version>1.0.0</version>
			<scope>provided</scope>
		</dependency>
```
And they're definitely all in there.
```
-rw-r--r-- 1 mark mark  106072 Apr  4 07:56 engine.io-client-1.0.0.jar
-rw-r--r-- 1 mark mark   64597 Apr  4 07:56 json-20180813.jar
-rw-r--r-- 1 mark mark  422787 Apr  4 07:56 okhttp-3.12.1.jar
-rw-r--r-- 1 mark mark   88732 Apr  4 07:56 okio-1.15.0.jar
-rw-r--r-- 1 mark mark   61944 Apr  4 16:14 org.openhab.binding.ambientweather-2.5.0-SNAPSHOT.jar
-rw-rw-r-- 1 mark mark      70 Apr 30  2017 README
-rw-r--r-- 1 mark mark   63004 Apr  4 07:56 socket.io-client-1.0.0.jar
```

davidgraeff commented:
Is the log saying anything interesting? (After a restart?) I'm not sure how the addons dir exactly works, if it maybe only processes OSGi bundles. But actually I don't think so, because I remember I had this situation once, but with bundles build by the old buildsystem.

The classpath file should be a copy of one of the already migrated bundles to be sure.

I image what would work is if you log into the karaf console (openhab-cli) and use bundle:install for each jar file

mhilbush commented:
After a restart, just this.
```
2019-04-04 16:13:07.896 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/opt/openhab2/addons/org.openhab.binding.ambientweather-2.5.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.binding.ambientweather [201]
  Unresolved requirement: Import-Package: io.socket.client

        at org.eclipse.osgi.container.Module.start(Module.java:444) ~[?:?]
        at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:383) ~[?:?]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260) [10:org.apache.felix.fileinstall:3.6.4]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233) [10:org.apache.felix.fileinstall:3.6.4]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startAllBundles(DirectoryWatcher.java:1221) [10:org.apache.felix.fileinstall:3.6.4]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:515) [10:org.apache.felix.fileinstall:3.6.4]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365) [10:org.apache.felix.fileinstall:3.6.4]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316) [10:org.apache.felix.fileinstall:3.6.4]
```
And, in the console, this.
```
openhab> bundle:stop org.openhab.binding.ambientweather                                                                   
openhab> bundle:start org.openhab.binding.ambientweather                                                                   
Error executing command: Error executing command on bundles:
	Error starting bundle 201: Could not resolve module: org.openhab.binding.ambientweather [201]
  Unresolved requirement: Import-Package: io.socket.client
```
Classpath looks like this
```
<?xml version="1.0" encoding="UTF-8"?>
<classpath>
	<classpathentry kind="src" output="target/classes" path="src/main/java">
		<attributes>
			<attribute name="optional" value="true" />
			<attribute name="maven.pomderived" value="true" />
		</attributes>
	</classpathentry>
	<classpathentry excluding="**" kind="src" output="target/classes" path="src/main/resources">
		<attributes>
			<attribute name="maven.pomderived" value="true" />
		</attributes>
	</classpathentry>
	<classpathentry kind="src" output="target/test-classes" path="src/test/java">
		<attributes>
			<attribute name="optional" value="true" />
			<attribute name="maven.pomderived" value="true" />
			<attribute name="test" value="true" />
		</attributes>
	</classpathentry>
	<classpathentry kind="con"
		path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8">
		<attributes>
			<attribute name="maven.pomderived" value="true" />
		</attributes>
	</classpathentry>
	<classpathentry kind="con" path="org.eclipse.m2e.MAVEN2_CLASSPATH_CONTAINER">
		<attributes>
			<attribute name="maven.pomderived" value="true" />
		</attributes>
	</classpathentry>
	<classpathentry kind="output" path="target/classes" />
</classpath>
```

I'm not sure what to make of this.
```
openhab> bundle:install file:///opt/openhab2/addons/socket.io-client-1.0.0.jar                                               
Bundle IDs: 
Error executing command: Error installing bundles:
	Unable to install bundle file:///opt/openhab2/addons/socket.io-client-1.0.0.jar: org.osgi.framework.BundleException: OSGi R3 bundle not supported
```

davidgraeff commented:
Can you try `bundle:install wrap:file:///opt/openhab2/addons/socket.io-client-1.0.0.jar` or even better use the maven coordinates `bundle:install wrap:mvn:commons-lang/commons-lang/2.4` (example)

mhilbush commented:
> bundle:install wrap:file:///opt/openhab2/addons/socket.io-client-1.0.0.jar
```
Error executing command: Error installing bundles:
	Unable to install bundle wrap:file:///opt/openhab2/addons/socket.io-client-1.0.0.jar: org.osgi.framework.BundleException: Error reading bundle content.
```

> bundle:install wrap:mvn:commons-lang/commons-lang/2.4

The format of this one is `groupId/artifactId/version`, so it should be this
```
bundle:install wrap:mvn:io.socket/socket.io-client/1.0.0
```
```
Error executing command: Error installing bundles:
	Unable to install bundle wrap:mvn:io.socket/socket.io-client/1.0.0: org.osgi.framework.BundleException: Error reading bundle content.
```

Just to check installing it directly from maven central...
```
bundle:install http://central.maven.org/maven2/io/socket/socket.io-client/1.0.0/socket.io-client-1.0.0.jar
Bundle IDs: 
Error executing command: Error installing bundles:
	Unable to install bundle http://central.maven.org/maven2/io/socket/socket.io-client/1.0.0/socket.io-client-1.0.0.jar: org.osgi.framework.BundleException: OSGi R3 bundle not supported
```

mhilbush commented:
I'm going to try building another binding that was migrated to use dependencies. I'll try amazondashbutton. If that works, then it's something I'm doing wrong.

mhilbush commented:
Ok. I tried amazondashbutton and I get the same issue. Even when I copy one of that binding's deps into addons -- in this case `pcap4j-core-1.6.6.jar`.
```
openhab> bundle:start org.openhab.binding.amazondashbutton                                                                 
Error executing command: Error executing command on bundles:
	Error starting bundle 269: Could not resolve module: org.openhab.binding.amazondashbutton [269]
  Unresolved requirement: Import-Package: org.pcap4j.core
```

davidgraeff commented:
With bundle:install you only provide maven coordinates not a web address. There's another command to load from a web address but I forgot.

mhilbush commented:
> With bundle:install you only provide maven coordinates

Ok, but wouldn't that be this, which I tried.
```
bundle:install mvn:io.socket/socket.io-client/1.0.0
```
or
```
bundle:install wrap:mvn:io.socket/socket.io-client/1.0.0
```

mhilbush commented:
@davidgraeff 

The only way I was able to get the binding to work in the addons directory was to create my own osgi bundle containing the 5 dependent jars, then drop that bundle into the addons directory.

I've never written one of these before, and it's throwing a few warnings when I run `mvn package`. But, when I drop the jar into addons, the dependencies are resolved and the binding runs.

Here's the pom.xml. This seems like a lot of trouble to be able to run a binding out of the addons directory. Isn't there a better way?
```
<project
  xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">

    <modelVersion>4.0.0</modelVersion>
  
    <groupId>com.hilbush</groupId>
    <artifactId>ambient-weather-deps</artifactId>
    <name>${project.artifactId} bundle</name>
    <description>Socket.io libs needed for Ambient Weather binding</description>
    <version>1.0.0</version>
    <packaging>bundle</packaging>

	<dependencies>
		<!-- https://mvnrepository.com/artifact/io.socket/socket.io-client -->
		<dependency>
			<groupId>io.socket</groupId>
			<artifactId>socket.io-client</artifactId>
			<version>1.0.0</version>
		</dependency>

		<!-- https://mvnrepository.com/artifact/com.squareup.okhttp3/okhttp -->
		<dependency>
			<groupId>com.squareup.okhttp3</groupId>
			<artifactId>okhttp</artifactId>
			<version>3.12.1</version>
		</dependency>

		<!-- https://mvnrepository.com/artifact/com.squareup.okio/okio -->
		<dependency>
			<groupId>com.squareup.okio</groupId>
			<artifactId>okio</artifactId>
			<version>1.15.0</version>
		</dependency>

		<!-- https://mvnrepository.com/artifact/io.socket/engine.io-client -->
		<dependency>
			<groupId>io.socket</groupId>
			<artifactId>engine.io-client</artifactId>
			<version>1.0.0</version>
		</dependency>

		<!-- https://mvnrepository.com/artifact/org.json/json -->
		<dependency>
			<groupId>org.json</groupId>
			<artifactId>json</artifactId>
			<version>20180813</version>
		</dependency>
	</dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>org.apache.felix</groupId>
        <artifactId>maven-bundle-plugin</artifactId>
        <version>4.0.0</version>
        <extensions>true</extensions>
        <configuration>
          <instructions>
            <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
            <Import-Package>
                android.*;resolution:=optional,
                com.android.org.*;resolution:=optional,
                dalvik.*;resolution:=optional,
                javax.annotation.meta.*;resolution:=optional,
                org.apache.harmony.*;resolution:=optional,
                org.conscrypt.*;resolution:=optional,
                sun.security.*;resolution:=optional
            </Import-Package>
            <Export-Package>
                io.socket.client,
                io.socket.emitter,
                org.json
            </Export-Package>
            <Embed-Dependency>*;scope=compile;inline=true</Embed-Dependency>
          </instructions>
        </configuration>
      </plugin>
    </plugins>
  </build>

</project>
```

davidgraeff commented:
Yes there is. Building a kar file. If #5222 is solved, I hope that can be tackled next. A kar file contains all dependency jars and a karaf feature.xml file and can be directly used by karaf using "feature:install" in the console or hopefully also via dropping those files into the addons directory.

But at the moment there is no better way, but there is not even a documentation written for developing with that new buildsystem.

J-N-K commented:
You can also use `bundle:install http://...` with the full URL (search the bundle in maven central, select the version and copy the link to the .jar). However, that only works for OSGI bundles, wrap is not working there (at least for me). 

Once the binding is merged and the feature.xml is deployed, the dependencies are automatically installed (at least unless you change the dependencies in a new version).

mhilbush commented:
Right. I understand that's the case once it's merged. It's really an issue when you want people to test a new version, or a brand new binding.

I also tried the bundle:install with the full URL. Definitely a no-go if it's not an osgi bundle.

mhilbush commented:
@davidgraeff I pushed several commits that address the feedback from your latest review.

And, now that I have a way to run the jar from the addons directory, I also removed the local 3rd party jars and specified those jars in the project pom.xml, and in feature.xml. I used the wrap: protocol in feature.xml since those jars are not osgi bundles. Hopefully, I did that correctly.

So, this is ready for review when you get a moment. Thanks!

As an aside, until there's a more systemic way to include non-osgi 3rd party jars when running a binding from addons, do you think I should post something on the forum explaining how I did it? It might save others from having to discover how to do it.

davidgraeff commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherEventListener.java:
```diff
@@ -0,0 +1,352 @@
+            options.transports = new String[] { "websocket" };
+            socket = IO.socket(url, options);
+        } catch (URISyntaxException e) {
+            logger.error("Listener: URISyntaxException getting IO socket: {}", e.getMessage());
```
Please do not use logger error level. This is reserved for the framework. Thanks.

mhilbush replied:
Good spot! Can't believe I missed that. Fixed now.

mhilbush commented:
@davidgraeff I pushed that last change to reduce the log level.

mhilbush commented:
In the latest commit, I reworked the processing of the weather event data returned from Ambient Weather. Occasionally, not all the weather data elements are returned (seems related to weather station batteries starting to get low, but there may be other causes). This had the effect, for example, of causing the temperature to fluctuate between the real temperature and 0 degrees, which tended to wreak havoc with rules. In the process of handling this situation better, I also reworked the code to make it a bit easier easier to add new weather station types in the future.

martinvw commented on bom/openhab-addons/pom.xml:
```diff
@@ -50,6 +50,11 @@
     </dependency>
+    <dependency>
+      <groupId>org.openhab.addons.bundles</groupId>
+      <artifactId>org.openhab.binding.ambientweather</artifactId>
```
alphabet wise a little bit of :-)

mhilbush replied:
Duh! And English is my native language. LOL

martinvw commented on bundles/org.openhab.binding.ambientweather/pom.xml:
```diff
@@ -0,0 +1,58 @@
+			<scope>provided</scope>
+		</dependency>
+        <!-- https://mvnrepository.com/artifact/io.socket/engine.io-client -->
+        <dependency>
```
Please format the pom.xml with spaces (its a mix now)

mhilbush replied:
Spaces or tabs? The Eclipse formatter wants to make it all tabs.

mhilbush replied:
Hmm. I see the POMs have been changed to spaces. Must be time to update my IDE. I've been dreading that day... 😦

martinvw commented on bundles/org.openhab.binding.ambientweather/pom.xml:
```diff
@@ -0,0 +1,58 @@
+        </dependency>
+		<!-- https://mvnrepository.com/artifact/com.squareup.okhttp3/okhttp -->
+		<dependency>
+			<groupId>com.squareup.okhttp3</groupId>
```
Is there a reason not to work with the standard HttpUtil or just plan Jetty? And any reason not to use gson?

martinvw replied:
I just saw your chat with @davidgraeff would be nice if you could come back to this binding later to switch to the defaults.

mhilbush replied:
I'll investigate if I can get Jetty Websockets to work after this gets merged. I think we're on a newer version of Jetty than when I first tried it.

martinvw commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,233 @@
+            logger.debug("Validating application and API keys");
+
+            String response = null;
+            try {
```
You could just join the code of two try-catches so that structure is clearer, you won't need the returns you can join the response declaration with the assignment.

mhilbush replied:
Yes, much simpler indeed.

martinvw commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,233 @@
+    /*
+     * Check if an API key has been provided in the thing config
+     */
+    private boolean haveApiKey() {
```
Nitpicking: hasApiKey?

mhilbush replied:
Yes to the nitpicking comment. 😄  Changed both method names.

martinvw commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherBridgeHandler.java:
```diff
@@ -0,0 +1,233 @@
+        return true;
+    }
+
+    public void updateThingStatus(String errorDetail, String statusDescription) {
```
Could you make this name more specific the application is at least way more specific.

mhilbush replied:
Done

martinvw commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/processor/ProcessorInterface.java:
```diff
@@ -0,0 +1,45 @@
+    /*
+     * Set the channel group Id for the station
+     */
+    public abstract void setChannelGroupId();
```
`public abstract` is implicit for methods declared in interfaces, better leave it out.

mhilbush replied:
Done

martinvw commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/model/EventSubscribedJson.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public boolean isMethodSubscribe() {
+        return "subscribe".equals(method) ? true : false;
```
please leave out ` ? true : false` it will do the same with less distraction

mhilbush replied:
Yep. Good catch.

martinvw commented on bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherEventListener.java:
```diff
@@ -0,0 +1,352 @@
+            }
+        } catch (JsonSyntaxException e) {
+            logger.info("Listener: Exception parsing subscribed event: {}", e.getMessage());
+            return;
```
You have a lot of return's like this could you sweep some of them out :-) also for example on line 297

mhilbush replied:
Swept.

mhilbush commented:
Thanks for your feedback @martinvw. I'll work on these over the next couple days.

As for the use of socket.io, I was unable to get Jetty websockets to work with the Ambient Weather back-end. When I switched to socket.io, it worked immediately. Of course, it's possible that I was not using the Jetty websocket API correctly. OTOH, I spent quite a few hours trying to get it to work. After I switched to socket.io, I did maintain consistency with the libs that openhab cloud was using, rather than introduce new versions.

mhilbush commented:
@martinvw I believe I've addressed all your review comments.

martinvw commented:
Thanks!

mhilbush commented:
@davidgraeff @martinvw Thanks for your time helping to get this merged!! I know you guys have been so busy with the migration.
